standards/SPIFFE-ID.md

```diff
@@ -74,6 +74,10 @@ Paths MAY be hierarchical - similar to filesystem paths. The specific meaning of
 
   ```spiffe://example.com/9eebccd2-12bf-40a6-b262-65fe0487d453```
 
+### 2.3. Maximum SPIFFE ID Length
+
+URIs, as defined by [RFC 3986](https://tools.ietf.org/html/rfc3986), do not have a maximal length. As an interoperability consideration, SPIFFE implementations MUST support SPIFFE URIs of at least 2048 characters in length, but MAY support longer. All URI components contribute to the URI length, including the "spiffe" scheme, "://" separator, trust domain, and path component. Note that [RFC 3986](https://tools.ietf.org/html/rfc3986) defines a maximum length of 255 characters for the "host" component of a URI; therefore a maximum length of a trust domain is 255 characters.
+
```

Comment:
Missing table of contents update, but otherwise LGTM

Reply:
Good catch. Fixed.

Reply:
> at least 2048 characters

Please define this limit in terms of bytes or define more exactly what a character is.

Reply:
I was probably influenced by RFC 3986's usage of 'characters'. But I agree with you that 'bytes' is clearer. Fixed.

Reply:
Thinking about this more, this characters vs bytes thing is a little more complex, and I do wonder whether we need to be specific about some points to ensure interop.

In RFC 3986, a URI is defined as a sequence of characters, and it explicitly avoids defining how that sequence of characters is encoded into bytes/octets (https://tools.ietf.org/html/rfc3986#section-2)

So nominally, if we are going to express the max length in terms of bytes, we need to be clear about the encoding used, e.g. UTF-8, ASCII, etc.

Also, URIs have percent-encoded characters in them, so it would be good to be explicit about how percent-encoded characters contribute to the length (this should be non-controversial if the length is in bytes and an encoding is specified).

One way to avoid this difficulty is to go back to talking about characters, but I think that's less useful to implementers who are trying to figure out how to size database columns, which was the original impetus for the clarification.

Reply:
Would your concerns be addressed if we clarified that the byte limit refers to the number of octets needed to store the URI, regardless of character encoding?

Since all of the characters in a URI contribute to it's length, I don't see why we need to be explicit about percent-encoded characters. My reasoning is that, just like character encoding, percent-encoded characters are interpreted after reading the bytes/octets which constitute the URI. And our goal here is to provide an upper-bound on the number of bytes/octets needed to store or transmit a URI.

Reply:
> the byte limit refers to the number of octets needed to store the URI, regardless of character encoding

That's just the problem. Without specifying the encoding it is not possible to count the number of octets required to store the URI. As a concrete example, consider a simple, boring URI with no funny business: spiffe://example.com/spike

In UTF-8, this takes 26 octets to encode.  In UTF-16, this takes 52 octets, since each character in the example takes 16 bits.

Reply:
> That's just the problem. Without specifying the encoding it is not possible to count the number of octets required to store the URI.

The encoding is specified by the fact that a URI must be a `IA5String`, an 8-bit ASCII encoding. See https://tools.ietf.org/html/rfc5280#section-7.4.

Reply:
@justinburke @evan2645 I've done some more reading of RFC 3986.

While a URI is defined as a sequence of _characters_, the set of allowed characters is restricted to Latin letters, digits, and a few symbols. This has the practical effect that a URI can always be successfully encoded at a rate of one octet per character using either US-ASCII or UTF-8 (these encodings are equivalent over the set of allowed characters).

Now, it may be the case that someone wants to generate a URI based on components that include a broader set of characters. They can do that, but the resulting URI will have all the characters outside the restricted set percent encoded. We only need to worry about the details of how that works if we need to interpret those components in some way more complicated than testing for equivalence. (For example, if we were ever going to propose matching semantics.)

So, the upshot of all this is that I think we can define the maximum length of a SPIFFE Identity as 2048 characters, including all components (scheme, separator, trust domain, path) rendered as a RFC3986 compliant URI. Defining the max length as a number of _characters_ fits the RFC definition, but should pose no difficulty to implementors, since they can encode any SPIFFE ID in 2048 or fewer _bytes_ using UTF-8.  We can mention this as "advice to implementors" to avoid confusion if people want.

Reply:
@briansmith that document describes how to encode a URI SAN in an x509 certificate.  A SPIFFE Identity is used in all kinds of contexts beyond x509. For example, JWTs, log files, configuration systems, etc.

A URI is not always a URI SAN.

Reply:
I hope it is true that a SPIFFIE ID is defined to be a URI and not an IRI, in which case bytes == characters because URIs use ASCII characters.

Anyway, what I really care about is that it is clear that in X.509, the maximum length is 2048 bytes, because an X.509 parser works in terms of bytes, and an X.509-based system is what I've built. The consequence of this is that maximum length would be 2048 characters in any realistic encoding.

Reply:
This was discussed during the SPIFFE SIG-Spec call on May 16th. The consensus at that time was that we should define the limit in terms of characters because this is the language that RFC 3986 uses, however we should be clear that (as @briansmith points out) URIs use ASCII characters and therefor the limit is effectively 2048 bytes. To be consistent with RFC 3986 terminology, as well as be explicit about what this means in terms of bytes, I am proposing the following text:

> URIs, as defined by [RFC 3986](https://tools.ietf.org/html/rfc3986), do not have a maximal length. As an interoperability consideration, SPIFFE implementations MUST support SPIFFE URIs up to 2048 characters in length and SHOULD NOT generate URIs of length greater than 2048 characters. [RFC 3986](https://tools.ietf.org/html/rfc3986) permits only ASCII characters, thus the maximum length of a SPIFFE ID is 2048 bytes.
> 
> All URI components contribute to the URI length, including the "spiffe" scheme, "://" separator, trust domain, and path component. Non-ASCII characters contribute to the URI length after they are percent encoded as ASCII characters. Note that [RFC 3986](https://tools.ietf.org/html/rfc3986) defines a maximum length of 255 characters for the "host" component of a URI; therefore a maximum length of a trust domain is 255 bytes.

Does this text address both of your concerns @briansmith @spikecurtis ?